ocaml/gencert/selfcert.ml

```diff
@@ -0,0 +1,165 @@
+(*
+ * Copyright (C) Citrix Systems Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published
+ * by the Free Software Foundation; version 2.1 only. with the special
+ * exception on linking described in file LICENSE.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *)
+
+module C = Cmdliner
+module Rsa = Mirage_crypto_pk.Rsa
+module UX = Xapi_stdext_unix.Unixext
+open Rresult (* introduces >>= >>| and R *)
+
+module D = Debug.Make (struct let name = "gencert_selfcert" end)
+
+(** initialize the random number generator at program startup when this
+module is loaded. *)
+let () = Mirage_crypto_rng_unix.initialize ()
+
+let defer f = Fun.protect ~finally:f
+
+(** [write_cert] writes a PEM file to [path]. It attempts to do that
+  atomically by writing to a temporary file in the same directory first and
+  renaming the file at the end *)
+let write_certs path key cert =
+  let delim = "\n" in
+  let temp_dir = Filename.dirname path in
+  let tmp = Filename.temp_file ~temp_dir "certify-" ".tmp" in
+  try
+    let fd = Unix.openfile tmp [Unix.O_WRONLY] 0o600 in
+    defer (fun () -> Unix.close fd) @@ fun () ->
+    UX.really_write fd (Cstruct.to_string key) 0 (Cstruct.len key) ;
+    UX.really_write_string fd delim ;
+    UX.really_write fd (Cstruct.to_string cert) 0 (Cstruct.len cert) ;
+    Unix.rename tmp path ;
+    R.ok ()
+  with e -> R.error_msg (Printexc.to_string e)
+
+let expire_in days =
+  let seconds = days * 24 * 60 * 60 in
+  let start = Ptime_clock.now () in
+  match Ptime.(add_span start @@ Span.of_int_s seconds) with
+  | Some expire ->
+      R.ok (start, expire)
+  | None ->
+      R.error_msgf "can't represent %d as time span" days
+
+let add_dns_names extension = function
+  | [] ->
+      extension
+  | names ->
+      X509.Extension.(
+        add Subject_alt_name
+          (false, X509.General_name.(singleton DNS names))
```

Comment:
Do you have any information on what the `false` means here, X509 documentation was not helpful when I checked.

Reply:
I don't. I looked at the code here but I agree that it is not obvious. The generated certs look correct to me:

https://github.com/yomimono/ocaml-certify/blob/primary/src/common.ml#L22-L27

Reply:
The parameter identifies "critical" extensions and I believe this parameter is not used here.